djangosaml2/views.py

```diff
@@ -254,104 +257,146 @@ def login(request,
+        except ResponseLifetimeExceed as e:
+            logger.info("SAML Assertion is no longer valid. Possibly caused by network delay or replay attack.", exc_info=True)
+            return fail_acs_response(request, exception=e)
+        except SignatureError as e:
+            logger.info("Invalid or malformed SAML Assertion.", exc_info=True)
+            return fail_acs_response(request, exception=e)
+        except StatusAuthnFailed as e:
+            logger.info("Authentication denied for user by IdP.", exc_info=True)
+            return fail_acs_response(request, exception=e)
+        except StatusRequestDenied as e:
+            logger.warning("Authentication interrupted at IdP.", exc_info=True)
+            return fail_acs_response(request, exception=e)
+        except StatusNoAuthnContext as e:
+            logger.warning("Missing Authentication Context from IdP.", exc_info=True)
+            return fail_acs_response(request, exception=e)
+        except MissingKey as e:
+            logger.exception("SAML Identity Provider is not configured correctly: certificate key is missing!")
+            return fail_acs_response(request, exception=e)
+        except UnsolicitedResponse as e:
+            logger.exception("Received SAMLResponse when no request has been made.")
+            return fail_acs_response(request, exception=e)
+
+        if response is None:
+            logger.warning("Invalid SAML Assertion received (unknown error).")
+            return fail_acs_response(request, status=400, exception=SuspiciousOperation('Unknown SAML2 error'))
+
+        session_id = response.session_id()
+        oq_cache.delete(session_id)
+
+        # authenticate the remote user
+        session_info = response.session_info()
+
+        if callable(attribute_mapping):
+            attribute_mapping = attribute_mapping()
+        if callable(create_unknown_user):
+            create_unknown_user = create_unknown_user()
+
+        logger.debug('Trying to authenticate the user. Session info: %s', session_info)
+        user = auth.authenticate(request=request,
+                                 session_info=session_info,
+                                 attribute_mapping=attribute_mapping,
+                                 create_unknown_user=create_unknown_user)
+        if user is None:
+            logger.warning("Could not authenticate user received in SAML Assertion. Session info: %s", session_info)
+            return fail_acs_response(request, exception=PermissionDenied('No user could be authenticated.'))
+
+        auth.login(self.request, user)
+        _set_subject_id(self.request.session, session_info['name_id'])
+        logger.debug("User %s authenticated via SSO.", user)
+        logger.debug('Sending the post_authenticated signal')
+
+        post_authenticated.send_robust(sender=user, session_info=session_info)
+        self.customize_session(user, session_info)
+
+        relay_state = self.build_relay_state()
```

Comment:
Could the `build_relay_state`, `custom_redirect ` and the `is_safe_url` be a single method? Which takes the (request, user, session_info) and possibly whatever else is useful (e.g. whether or not to perform the is_safe_check, this might want to be customized entirely), an returns a 'definitive' url to redirect to. And then the default implementation does what is currently done. That way it's a bit clearer due there being a single method to determine the redirect, which users can override, instead of the several pieces which all do a part of it.